New TokenPaymaster

forshtat opened the pull request.

forshtat committed:
Bring over PimlicoERC20Paymaster as new TokenPaymaster

forshtat committed:
Pass local lint and original tests

forshtat committed:
Fix bug in denominators calculation; streamline tests; use OZ SafeERC20

forshtat committed:
Reuse 'updatePrice' logic and test 'TokenPriceUpdated'

forshtat committed:
Extract oracle-specific logic to the OracleHelper library

forshtat committed:
Bring over necessary UniswapHelper code - solidity compiles

forshtat committed:
Some cleanup

forshtat committed:
Deploy TokenPaymaster witout Create2 factory for the unit tests

forshtat committed:
REPRO: out of gas in 'handleOps'

forshtat committed:
Add 'TestUniswap' to silence errors; fix price-related errors

forshtat committed:
Add token price override and test for it

forshtat committed:
WIP: Implemented TestUniswap stub logic and deposit to EntryPoint logic

forshtat committed:
Remove all 'console.log'

forshtat committed:
Fix accidental change in EP

huaweigu commented on contracts/samples/TokenPaymaster.sol:
minor, IIRC from our past debug experience, SafeERC20 uses banned [selfbalance opcode](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/dfef6a68ee18dbd2e1f5a099061a3b8a0e404485/contracts/utils/Address.sol#L134)

forshtat replied:
Thanks, I did not know that! Created a small PR in the OpenZeppelin repo, hope this is all there is to this issue:
https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4271

I hope to avoid having to copy `SafeERC20.sol` library into the `account-abstraction` repo, but maybe we will have no choice.

leekt replied:
maybe we can use solady in that case https://github.com/Vectorized/solady/blob/f9a6cc652dc8b37387590588a78644abfa4be3ac/src/utils/SafeTransferLib.sol#L238

huaweigu commented on contracts/samples/TokenPaymaster.sol:
I wonder if it's worth of attempting to transfer the owed token (preCharge - actualTokenNeeded) from sender here

forshtat replied:
I think this may actually work, will bring this up with the team. So far I did add an attempt to withdraw the owed amount (`actualTokenNeeded - preCharge`) from the UserOp sender to the Paymaster.

huaweigu commented on contracts/samples/TokenPaymaster.sol:
Out of curiosity, how do we pick the upper bound (1900000)?

forshtat replied:
None of these hard-coded parameters are not yet selected, but in my opinion the maximum possible value for the markup should just be there to prevent human error or malicious configuration.

For example, Pimlico Paymaster used `120e4` which sounds to me like it may be a bit too limiting:
https://github.com/pimlicolabs/erc20-paymaster-contracts/blob/c9911402a49928398441b07fb9caa9f4b4d0352f/src/PimlicoERC20Paymaster.sol#L68

livingrockrises replied:
maybe 2e6 is good upper limit

forshtat replied:
@livingrockrises Agreed. Will set for 200% for now. Will also make it explicitly related to the `PRICE_DENOMINATOR` value.

huaweigu commented on contracts/samples/TokenPaymaster.sol:
nitpick: maybe we shouldn't revert if preCharge == actualTokenNeeded (edge case)

forshtat replied:
Right, thanks, will fix. To be clear, there is almost zero chance of that happening.

forshtat committed:
TBD: charge the sender account in postOp if 'refund' is negative ('overdraft')

livingrockrises commented on contracts/samples/TokenPaymaster.sol:
Does this mean router swapping cost is dumped on to the users? 
since it is conditional postOp cost would change 
if not user who is paying for this computation - paymaster's gas deposit?

forshtat replied:
A single user should not get stuck with the bill for the Paymaster's swap operation on its own.
I am not sure but I think what we should do is to set the `POSTOP_COST` parameter such that it kind of spreads the cost of these periodic swaps over all other transactions.
Of course this way it may end up accumulating some loss or profit over time if left completely unattended.

livingrockrises replied:
If paymaster is always charging premium, POSTOP_COST can just be everything unaccounted on-chain on entry point and postOp - swap executionGas.

drortirosh replied:
@livingrockrises 
A paymaster is expected to set `refundPostopCost` to actual cost of postOp, with faction of the cost of the uniswap, since the swap is expected to happen only every "N" transaction.

forshtat committed:
WIP: Check new price calculation function; use 1e26 as 'price denominator'

leekt commented on contracts/samples/TokenPaymaster.sol:
STATICCALL to entrypoint is not allowed
```
Limitation on “CALL” opcodes (CALL, DELEGATECALL, CALLCODE, STATICCALL):
must not use value (except from account to the entrypoint)
must not revert with out-of-gas
destination address must have code (EXTCODESIZE>0)
cannot call EntryPoint’s methods, except depositFor (to avoid recursion)
```

livingrockrises commented on contracts/samples/TokenPaymaster.sol:
instead of using maxFeePerGas, does it make sense to use tx.gasprice ?

forshtat replied:
You are right, thanks! However, we cannot just use `tx.gasprice` as the bundler can manipulate this value. I think we should use `userOp.gasPrice()` here. Updated.

livingrockrises replied:
userop.gasPrice() touches BASEFEE which forbidden opcode. this would be fine in postOp(), but you shouldn't calculate it in validatePaymasterUserOp and pass it in context.
Instead pass maxPriorityFee and maxFee in context and calculate like below in the postOp

```
uint256 opGasPrice;
        unchecked {
            if (maxFeePerGas == maxPriorityFeePerGas) {
                opGasPrice = maxFeePerGas;
            } else {
                opGasPrice = Math.min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
            }
        }
```

forshtat replied:
Right 😀

livingrockrises commented on contracts/samples/TokenPaymaster.sol:
what is the rationale behind using 1e26?

forshtat replied:
At this point it constants are actually pretty random and will be finalised after we make some gas usage optimisations. Do you have suggestions on what value should be used and why?

forshtat committed:
Use 'userOp.gasPrice' instead of 'maxFeePerGas' to calculate precharge/refund

forshtat committed:
Avoid duplicate return from '_postOp'

forshtat commented on contracts/samples/TokenPaymaster.sol:
Thanks. Fixed.

livingrockrises commented on contracts/samples/TokenPaymaster.sol:
this will revert on the bundler
cc @leekt

leekt replied:
yup this will revert for sure

forshtat committed:
Correction: pre-charge based on 'maxFeePerGas', refund based on 'gasPrice'

This is the behaviour of the EntryPoint contract as well.

forshtat committed:
Add support for legacy chains (avoid 'basefee' opcode)

leekt commented on contracts/samples/TokenPaymaster.sol:
Not sure about this part. Won't this make `clientSuppliedPrice` invalid? `clientSuppliedPrice` should be the guard for how much token user is willing to pay but this part will end up forcing user to pay the full slippage

forshtat replied:
Well, yes, with this implementation the user-supplied price can only increase the pre-charge for the user, but does not set an upper limit of the token price, so once the users approve this paymaster they basically agree to pay for gas at any current market price. 

@drortirosh what do you think about this feature?

CC: @huaweigu

forshtat commented on contracts/samples/TokenPaymaster.sol:
This paymaster will accumulate client base by getting 'approval' in ERC-20 token used. Deploying a new paymaster just to switch to a different price oracle will lose all existing clients until they migrate their approval.
Do you want me to make the fields immutable and make the paymaster an upgradeable proxy instead?

forshtat replied:
@drortirosh
Bumping this one. Should we?

drortirosh replied:
As a user, I trust the config values the paymaster have - but I really can't fully trust it, since it can change its configuration at will  (e.g. increase the postOp cost, or switch to a malicious oracle)

I want to trust an address given by some authority, and know this address can't change.
Better yet - I want to know this address define all parameters values.

Also other parameters (like minEPbalance) are not expected to be modified much.

livingrockrises replied:
at Biconomy we are returning a fee quote to the sender and allow them to update userop.calldata with the approval transaction with necessary amount each time. then partialUserOp is filled with paymasterAndData and signature. SDK account APIs can facilitate this for decoding calldata and append approval transaction in a batch and have new callData, callGasLimit

forshtat committed:
Switch from 'abi.encodePacked' to 'abi.encode' for better readability

forshtat committed:
Make 'refundPostopCost' a part of 'TokenPaymasterConfig'

forshtat committed:
Return price invalidation time to prevent ops using against outdated price

forshtat committed:
Reorder lines of code to make the fast way out even more faster

forshtat commented on contracts/samples/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,147 @@
+        IOracle nativeOracle = oracleHelperConfig.nativeOracle;
+
+        uint256 cacheAge = block.timestamp - cachedPriceTimestamp;
+        if (!force && cacheAge <= cacheTimeToLive) {
```
Fixed.

forshtat commented on contracts/samples/TokenPaymaster.sol:
Done.

forshtat commented on contracts/samples/TokenPaymaster.sol:
```diff
@@ -1,113 +1,224 @@
+    uint256 private constant PRICE_DENOMINATOR = 1e26;
+
+    /// @notice Estimated gas cost for refunding tokens after the transaction is completed
+    uint256 public constant REFUND_POSTOP_COST = 40000;
```
Done.

drortirosh commented on contracts/samples/TokenPaymaster.sol:
```diff
@@ -142,8 +145,12 @@ contract TokenPaymaster is BasePaymaster, UniswapHelper, OracleHelper {
+            context = abi.encode(tokenAmount, userOp.maxFeePerGas, userOp.maxPriorityFeePerGas, userOp.sender);
+            validationResult = _packValidationData(
+                false,
+                uint48(cachedPriceTimestamp + tokenPaymasterConfig.priceMaxAge),
```
storage optimization: instead referencing 2 storage slots on each call, use 
`cachedPriceExpirationTimestamp = timestamp + priceMaxAge`

forshtat replied:
Please elaborate where does the `timestamp` variable come from?

drortirosh commented on test/samples/OracleHelper.test.ts:
```diff
@@ -102,6 +102,8 @@ describe.only('OracleHelper', function () {
     const owner = await ethersSigner.getAddress()
 
     const tokenPaymasterConfig: TokenPaymaster.TokenPaymasterConfigStruct = {
+      priceMaxAge: 86400,
```
^^ remove "only" from `describe.only('OracleHelper')`

drortirosh commented on contracts/index.ts:
```diff
@@ -1,2 +0,0 @@
-export { UserOperationStruct } from './types/EntryPoint'
```
true, should be removed (along with the code to use it during "yarn publish"), but unrelated to TokenPaymaster.

forshtat replied:
This was messing up the environment for me though.

drortirosh commented on contracts/test/TestUniswap.sol:
```diff
@@ -0,0 +1,60 @@
+
+    /// @notice Simplified code copied from here:
+    /// https://github.com/Uniswap/v3-periphery/blob/main/contracts/base/PeripheryPayments.sol#L19
+    function unwrapWETH9(uint256 amountMinimum, address recipient) public payable {
```
unused?

forshtat replied:
Why not? This is used by the `UniswapHelper.sol` file:

```
    function unwrapWeth(uint256 amount) internal {
        IPeripheryPayments(address(uniswap)).unwrapWETH9(amount, address(this));
    }
```

drortirosh commented on contracts/samples/TokenPaymaster.sol:
I think this is `_cachedPrice`, not `cachedPrice`

drortirosh reviewed (commented):
Should also add gas-check test in `gas-calc/`

forshtat committed:
Address PR comments (see list)

1. Remove 'setOracleConfiguration' - only constructor can set oracles

2. Use '_cachedPrice' instead of 'cachedPrice' to avoid extra state read

3. Remove 'only' from TokenPaymaster related tests

drortirosh commented on contracts/samples/utils/OracleHelper.sol:
```diff
@@ -67,7 +67,7 @@ abstract contract OracleHelper {
     function _setOracleConfiguration(
         OracleHelperConfig memory _oracleHelperConfig
-    ) internal {
+    ) private {
```
leaving this method (called only from constructor) means that we still pay the SLOAD when using these parameters.
(and you can't make a member "immutable" if it is set from a private method)

Currently, the paymaster uses 15 slots (14, if we exclude "owner")
(checked with `solc --storage-layout`)
So it uses 14 storage slots on each operation, which is at minimum 14*2100 = 29400 gas .
But I suggest adding a "gascalc" test before optimizing it.
